lib/vnet/ssh_proxy.go

```diff
@@ -0,0 +1,239 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// proxySSHConnection transparently proxies incoming SSH channels and requests
+// to a target SSH client.
+func proxySSHConnection(
+	ctx context.Context,
+	targetClient *ssh.Client,
+	channels <-chan ssh.NewChannel,
+	requests <-chan *ssh.Request,
+) error {
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		return proxyChannels(ctx, targetClient, channels)
+	})
+	g.Go(func() error {
+		return proxyGlobalRequests(ctx, targetClient, requests)
+	})
+	return trace.Wrap(g.Wait(), "proxying SSH connection")
+}
+
+func proxyGlobalRequests(
+	ctx context.Context,
+	targetClient *ssh.Client,
+	requests <-chan *ssh.Request,
+) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case req, ok := <-requests:
+			if !ok {
+				return nil
+			}
+			ok, reply, err := targetClient.SendRequest(req.Type, req.WantReply, req.Payload)
+			if err != nil {
+				if !req.WantReply {
+					// No reply was expected, log the error and continue
+					// handling global requests.
+					log.WarnContext(ctx, "Failed to forward global SSH request to target", "error", err)
+					continue
+				}
+				err = trace.Wrap(err, "forwarding global request to target")
+				if replyErr := req.Reply(false, nil); replyErr != nil {
+					// A reply was expected and we failed to send one, we're in a
+					// bad state, return an error to terminate the connection.
+					return trace.NewAggregate(err, trace.Wrap(replyErr, "replying to request with error"))
+				}
+				// We failed to send the request, but we informed the client of
+				// the failure with req.Reply(false, nil), so it's safe to
+				// continue.
+				continue
+			}
+			if err := req.Reply(ok, reply); err != nil {
+				// A reply was expected by the client and returned by the target
+				// but we failed to forward it back. We're in a bad state,
+				// return an error to terminate the connection.
+				return trace.Wrap(err, "forwarding reply from target back to client")
+			}
+		}
+	}
+}
+
+func proxyChannels(
+	ctx context.Context,
+	targetClient *ssh.Client,
+	channels <-chan ssh.NewChannel,
+) (err error) {
+	// Every channel will be handled in its own goroutine, but a recoverable
+	// error from a single channel should not terminate other channels, so
+	// handleNewChannel must avoid returning errors that should not terminate
+	// the connection.
```

Comment:
The connection being terminated will be detected by `Wait`, there's no reason why the channel handler returning an error should result in the connection being terminated at all.

Reply:
this is handled completely differently now